src/abstractmcmc.jl

```diff
@@ -303,3 +239,138 @@ function (cb::HMCProgressCallback)(rng, model, spl, t, state, i; nadapts = 0, kw
         @info "Finished $nadapts adapation steps" adaptor κ.τ.integrator metric
     end
 end
+
+#############
+### Utils ###
+#############
+
+function get_type_of_spl(spl::AbstractHMCSampler)
+    T = collect(typeof(spl).parameters)[1]
+    return T
+end
+
+const SYMBOL_TO_INTEGRATOR_TYPE = Dict(
+    :leapfrog => Leapfrog,
+    :jitterleapfro => JitteredLeapfrog,
+    :temperedleapfrog => TemperedLeapfrog,
+)
+
+function determine_integrator_constructor(integrator::Symbol)
+    if !haskey(SYMBOL_TO_INTEGRATOR_TYPE, integrator)
+        error("Integrator $integrator not supported.")
+    end
+
+    return SYMBOL_TO_INTEGRATOR_TYPE[integrator]
+end
+
+# If it's the "constructor" of an integrator or instantance of an integrator, do nothing.
+determine_integrator_constructor(x::AbstractIntegrator) = x
+determine_integrator_constructor(x::Type{<:AbstractIntegrator}) = x
+determine_integrator_constructor(x) = error("Integrator $x not supported.")
```

Comment:
Can we use a different design? It seems quite weird... Why do we need to retrieve the constructor? Couldn't we just use something like `get_integrator(Val(spl.integrator), eps)` where we dispatch on the first argument?

Reply:
Ok let me try!

Reply:
I think we still want to keep the look up table design but I have disentangled `make_step_size `from `make_integrator`.
@torfjelde?

Reply:
I like this solution a lot!! I have given the same treatment to the integrator and the metric. 
There's a little bit of nesting going on here but basically if the user provides an `AbstractMetric`/`AbstractIntegrator ` the code simly returns it. If on the other hand, the user provides a symbol it gets wrapped in  `Val(::Symbol)` and then we do your solution.

This allows the user to use the NUTS constructor with their own metric/integrator without having to build the whole sampler themselves which is something I always wanted to code into these constructors. 

```julia
make_integrator(spl::HMCSampler, ϵ::Real) = spl.κ.τ.integrator
make_integrator(spl::AbstractHMCSampler, ϵ::Real) = make_integrator(spl.integrator, ϵ)
make_integrator(i::AbstractIntegrator, ϵ::Real) = i
make_integrator(i::Type{<:AbstractIntegrator}, ϵ::Real) = i
make_integrator(i::Symbol, ϵ::Real) = make_integrator(Val(i), ϵ)
make_integrator(i...) = error("Integrator $(typeof(i)) not supported.")
make_integrator(i::Val{:leapfrog}, ϵ::Real) = Leapfrog(ϵ)
make_integrator(i::Val{:jitteredleapfrog}, ϵ::Real) = JitteredLeapfrog(ϵ)
make_integrator(i::Val{:temperedleapfrog}, ϵ::Real) = TemperedLeapfrog(ϵ)

#########

make_metric(i...) = error("Metric $(typeof(i)) not supported.")
make_metric(i::Symbol, T::Type, d::Int) = make_metric(Val(i), T, d)
make_metric(i::AbstractMetric, T::Type, d::Int) = i
make_metric(i::Type{AbstractMetric}, T::Type, d::Int) = i
make_metric(i::Val{:diagonal}, T::Type, d::Int) = DiagEuclideanMetric(T, d)
make_metric(i::Val{:unit}, T::Type, d::Int) = UnitEuclideanMetric(T, d)
make_metric(i::Val{:dense}, T::Type, d::Int) = DenseEuclideanMetric(T, d)
```